### PR TITLE
[utils] `traverse_obj`: Add `set` transformations/filters and `re.Match` group names

### PR DESCRIPTION
**Pull request by Grub4K**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Add transformation functions and type filters (through wrapping them into a `set`) and named groups for `re.Match` in filter functions to `traverse_obj`. It also adds a sanity check for filter functions in debug mode to require 2 parameters.
```py
traverse_obj(obj, ("some", "path", {int_or_none}))  # Convert to int
traverse_obj(obj, ("some", "path", {int}))  # Filter non int values
traverse_obj(re.findall("...", "..."), (..., lambda k, _: k.startswith("name")))
traverse_obj(..., lambda x: ...)  # Error in debug mode
```

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/utils.py [hunk 1]
```diff
@@ -5473,6 +5477,10 @@ def apply_key(key, obj):
         elif key is None:
             yield obj
 
+        elif isinstance(key, set):
+            assert len(key) == 1
```
Comment (pukkandan) [inline yt_dlp/utils.py:5481]:
```suggestion
            assert len(key) == 1, 'Set should only be used to wrap transformations'
```

File: yt_dlp/utils.py [hunk 2]
```diff
@@ -5551,6 +5561,9 @@ def apply_path(start_obj, path):
             if key is ... or isinstance(key, (list, tuple)) or callable(key):
                 has_branched = True
 
+            if __debug__ and callable(key):
+                inspect.signature(key).bind(None, None)
```
Comment (pukkandan) [inline yt_dlp/utils.py:5565]:
```suggestion
                inspect.signature(key).bind(None, None)  # Verify function signature
```

File: test/test_utils.py [hunk 3]
```diff
@@ -2015,6 +2015,27 @@ def test_traverse_obj(self):
                          msg='function as query key should perform a filter based on (key, value)')
         self.assertCountEqual(traverse_obj(_TEST_DATA, lambda _, x: isinstance(x[0], str)), {'str'},
                               msg='exceptions in the query function should be catched')
+        if __debug__:
```
Comment (pukkandan) [inline test/test_utils.py:2018]:
```suggestion

        # Test wrong function signature
        if __debug__:
```

**Review by pukkandan — APPROVED** [review]

File: test/test_utils.py (hunk 3 above)
Comment (Grub4K) [reply test/test_utils.py:2018]: How about embedding the comment in the assertion message?
```py
        if __debug__:
            with self.assertRaises(Exception, msg='Wrong function signature should raise in debug'):
                traverse_obj(_TEST_DATA, lambda a: ...)
            with self.assertRaises(Exception, msg='Wrong function signature should raise in debug'):
                traverse_obj(_TEST_DATA, lambda a, b, c: ...)
```

Comment (pukkandan) [reply test/test_utils.py:2018]: I suggested this coz you have comments for the other sections

Comment (Grub4K) [reply test/test_utils.py:2018]: But this is the same section (Query functions)?

**Pull request merged**